generators/add-vscode-config/index.js

```diff
@@ -0,0 +1,209 @@
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const Generator = require('yeoman-generator')
+const path = require('path')
+const fs = require('fs-extra')
+const { absApp, objGetValue } = require('./utils')
+
+const {
+  createVsCodeConfiguration,
+  createLaunchCompound,
+  createChromeLaunchConfiguration,
+  createPwaNodeLaunchConfiguration
+} = require('./VsCodeConfiguration')
+
+/*
+    'initializing',
+    'prompting',
+    'configuring',
+    'default',
+    'writing',
+    'conflicts',
+    'install',
+    'end'
+*/
+
+const Default = {
+  DESTINATION_FILE: '.vscode/launch.json',
+  REMOTE_ROOT: '/code'
+}
+
+const Option = {
+  DESTINATION_FILE: 'destination-file',
+  FRONTEND_URL: 'frontend-url',
+  REMOTE_ROOT: 'remote-root',
+  APP_CONFIG: 'app-config'
+}
+
+class AddVsCodeConfig extends Generator {
+  constructor (args, opts) {
+    super(args, opts)
+
+    // options are inputs from CLI or yeoman parent generator
+    this.option(Option.APP_CONFIG, { type: Object })
+    this.option(Option.FRONTEND_URL, { type: String })
+    this.option(Option.REMOTE_ROOT, { type: String, default: Default.REMOTE_ROOT })
+    this.option(Option.DESTINATION_FILE, { type: String, default: Default.DESTINATION_FILE })
+  }
+
+  verifyConfig () {
+    const appConfig = this.options[Option.APP_CONFIG]
+    const verifyKeys = [
+      'app.hasFrontend',
+      'app.hasBackend',
+      'ow.package',
+      'ow.apihost',
+      'manifest.package.actions',
+      'web.src',
+      'web.distDev',
+      'root',
+      'envFile'
+    ]
+
+    const missingKeys = []
+    verifyKeys.forEach(key => {
+      if (objGetValue(appConfig, key) === undefined) {
+        missingKeys.push(key)
+      }
+    })
+
+    if (missingKeys.length > 0) {
+      throw new Error(`App config missing keys: ${missingKeys.join(', ')}`)
+    }
+  }
+
+  _getActionEntryFile (pkgJson) {
+    const pkgJsonContent = fs.readJsonSync(pkgJson)
+    if (pkgJsonContent.main) {
+      return pkgJsonContent.main
+    }
+    return 'index.js'
+  }
+
+  _processRuntimeArgsForActionEntryFile (action, runtimeArgs) {
+    const appConfig = this.options[Option.APP_CONFIG]
+    const actionPath = absApp(appConfig.root, action.function)
+
+    const actionFileStats = fs.lstatSync(actionPath)
+    if (actionFileStats.isDirectory()) {
+      // take package.json main or 'index.js'
+      const zipMain = this._getActionEntryFile(path.join(actionPath, 'package.json'))
+      return path.join(actionPath, zipMain)
+    }
+
+    return runtimeArgs
+  }
+
+  _processForBackend () {
+    const appConfig = this.options[Option.APP_CONFIG]
+    const nodeVersion = this.options[Option.NODE_VERSION]
+    const remoteRoot = this.options[Option.REMOTE_ROOT]
+
+    const packageName = appConfig.ow.package
+    const manifestActions = appConfig.manifest.package.actions
```

Comment:
I know I'm commenting on my own PR, but just so I won't forget 😅  This won't work for custom package names (see the updated implementation in https://github.com/adobe/aio-cli-plugin-app/pull/343)

Reply:
resolved in c1ad15c